Envelope fix

neumatho opened the pull request:
Two fixes in this pull request:

First fix will prevent last used sampled in a channel to be played, if afterwards an illegal instrument is set. Now nothing is played.

The second fix is a big one. Almost totally rewrote the envelope handling, so it behave like FastTracker II does, specially when used together with XM effect L. Inspiration to this fix is from a translation of the original FastTracker II code to C, which can be found here: [Replayer source to FT II Clone](https://github.com/8bitbubsy/ft2-clone/blob/master/src/ft2_replayer.c). It is special the updateChannel() function I looked at.

These two fixes make Ebony Owl Netsuke.xm to play correctly.

[Ebony Owl Netsuke.zip](https://github.com/sezero/mikmod/files/7476168/Ebony.Owl.Netsuke.zip)

sezero commented:
@AliceLR please have a look when you can.

AliceLR commented:
> The second fix is a big one. Almost totally rewrote the envelope handling

Changes that affect multiple formats like this with no regression testing are, as sezero can confirm, a *significant* contributor to why MikMod's format accuracy has been in such a sad state for a long time. How many modules have you tested this patch against? How many are XM? IT? IMAGO Orpheus?

> First fix will prevent last used sampled in a channel to be played, if afterwards an illegal instrument is set. Now nothing is played.

I think this is another thing that can vary significantly between formats (maybe @sagamusix can confirm...).

sagamusix commented:
There are significant differences between IT and XM style envelope handling, so just taking an XM player's code here is almost guaranteed to break other formats unless they were already broken to begin with.

> I think this is another thing that can vary significantly between formats (maybe @sagamusix can confirm...).

I don't have tests handy but IIRC IT in instrument mode will disagree with that due to how the instrument to sample mapping works.

neumatho commented:
I tried to be very careful when doing this change, because I know it may broke something. Did a lot of debugging on multiple modules. I have tested with some modules, but not that many. Will test more. Ofcourse it would be good, if there is some real good testing modules in all formats where you can easily hear how it should sound and if it is broken.

I did not just copy the XM envelope function, but tried to apply what this did in some situation with the original MikMod code. The only difference I could see out of MikMod code between XM and IT, is that XM has one sustain point, while IT has two (begin/end). This part is kept in my changes. Are there other differences I need to be aware of?

And while writing this, there is one thing I need to check. I'm not sure if there is made interpolation when running between multiple sustain points. Will check that and commit any changes if needed.

neumatho commented:
It seems there are some issues with IT. The strange thing is that my C# port and the original MikMod code do not sound similar, which is very strange. I'm looking at it.

neumatho commented:
After many hours of analyse and debugging, I finally found why there are differences. It has nothing to do with this patch. In the module "A Life In Termoil.it", there is played a sequence in 4 voices at the beginning. In the original MikMod code, only the first 2 can be heard, even when the code parses them, while my port of MikMod all 4 can be heard. I then found out, that the last two voices are marked as surround channels (via S9x effect). If I disable surround (via DMODE flag), then all 4 channels are played in MikMod. So there seem to be an issue with surround mixing in MikMod.

Back to my patch. This patch does not sound correctly on IT modules, so I think I will make two separate envelope handlers for XM and IT. I then need to figure out exactly how it works. Do any of you know what IMAGO Orpheus uses? XM, IT or its own?

sagamusix commented:
In general no two trackers from that period are alike because nobody really wrote done in a fully detailed way how their engine works. However, Imago behaves more like FT2 than Impulse Tracker when it comes to envelope handling, so as long as mikmod doesn't have any specialized handling for envelopes in IMF yet, it's probably best to follow what XM does.

neumatho commented:
I have now separated the envelope handlers into two different methods. The main difference between the XM and IT way of handling envelopes, is that XM process the first point before it checks for sustain/looping, while IT do it for each tick. Lets take an example. We have a simple envelope of two points:

```
Pos  Value
 0    256
12      0
```

The loop has both a start and end on point 0. For XM, the volume will start at 256 and then fade down to 0 in 12 ticks and move to the next point. Because the new point is after the loop, the loop is ignored, so it keep the volume at 0.

For IT, it will keep the volume at 256 the whole time, because it keep looping at the first point.

For the illegal instrument fix, MikMod has always made a test for out-of-bounds instruments, it just forgot to clear some variables, which made it play the last used sample on that sample, whatever that was. So the fix, is just to clear those variables.

sezero commented:
Rebased this to current master. Reviews? Comments?

sezero commented:
@AliceLR: What is your final take in this? Reviews, comments?

neumatho commented:
Just wanna say, that libxmp has the same problems.

AliceLR commented on libmikmod/playercode/mplayer.c:
```diff
@@ -3015,13 +3120,18 @@ static void pt_Notes(MODULE *mod)
 				break;
 			case UNI_INSTRUMENT:
 				inst=UniGetByte();
-				if (inst>=mod->numins) break; /* safety valve */
+				if (inst>=mod->numins) {
+					a->main.i=NULL;
+					a->main.s=NULL;
+					a->main.sample=-1;
+					break; /* safety valve */
+				}
```
I don't have any useful examples right now but I *think* this might be completely format-dependent. Judging just on how libxmp does this (not necessarily correct, but generally pretty good):

* MOD: completely cuts the playing note.
* XM/other XM-likes: sets the volume to 0 but continues playing the current note.
* S3M/other S3M-likes: ignores the invalid instrument, continues playing the current note.
* IT (sample mode): sets the volume to 0 but continues playing the current note.
* IT (instrument mode): ignores the invalid instrument, continues playing the current note.
* MED: completely cuts the playing note.

If the above behavior descriptions are accurate, this change *could* be at the expense of S3M, XM, and IT accuracy.

AliceLR commented on libmikmod/playercode/mplayer.c:
```diff
@@ -2806,30 +2909,32 @@ static void pt_UpdateVoices(MODULE *mod, int max_volume)
-				envpan = StartEnvelope(&aout->penv,aout->main.panflg,
-				  i->panpts,i->pansusbeg,i->pansusend,
-				  i->panbeg,i->panend,i->panenv,aout->main.keyoff);
+			if (aout->main.volflg & EF_ON) {
```
The changes in this section look 100% correct, I would merely recommend staying close to the old text wrapping where possible (I really hope no one relies on 96-char wrapping though...).

AliceLR replied:
Never mind, the text wrapping in other parts of this file are already a mess.

AliceLR commented on libmikmod/playercode/mplayer.c:
```diff
@@ -1613,21 +1719,18 @@ static int DoXMEffectL(UWORD tick, UWORD flags, MP_CONTROL *a, MODULE *mod, SWOR
 	if ((!tick)&&(a->main.i)) {
 		INSTRUMENT *i=a->main.i;
-		MP_VOICE *aout;
+		MP_VOICE *aout = &mod->voice[channel];
```
Not sure what exactly is going on with MikMod's handling of channel assignment; is there a reason this was changed from `a->slave`?

neumatho replied:
It is a while ago, since I made this patch, so I cannot exactly remember. I think it is because a->slave is null, since it is a XM effect, so I need to get the voice in a different way.

AliceLR reviewed (commented):
So far I've reviewed everything but `ProcessEnvelope` and cohorts, and I don't think there are many issues of note. And yes, libxmp's envelope code leaves a bit to be desired, there's already an issue open for *part* of this I think :)

neumatho committed:
Envelope fixes [1/2]:

Prevent last used sampled in a channel to be played, if afterwards
an illegal instrument is set. Now nothing is played.

Combined with the next patch, these two fixes make Ebony Owl Netsuke.xm
to play correctly.

neumatho committed:
Envelope fixes [2/2]:

Updated the envelopes, so there are two separate handlers. One for
XM/IMF and one for IT.

Almost totally rewrote the envelope handling, so it behaves like
FastTracker II does, specially when used together with XM effect L.

These two fixes make Ebony Owl Netsuke.xm to play correctly.

sezero commented:
Rebased to mainstream with build fixes (no TRUE/FALSE..)

AliceLR commented:
Will rebase/patch this for #75 once it's in.

sezero commented:
> Will rebase/patch this for #75 once it's in.

PING?
